Monotype/c2pa-rs
# C2PA-147: General box hashes for fonts

# JIRA Issue

- C2PA-147

# Checklist


# Notes for Reviewers

Although general box hash support was added to the SDK, I found no code other than integration tests calling the new `AssetBoxHash` trait on asset IO handlers. 

With that said, the way chunks/positions were read from fonts were redone to use a trait for abstraction at a later point for other font formats.

One of the more controversial changes is to the `store.rs` module. I added the ability to add multiple exclusions for the data hashing, defragmenting locations when they were actually contiguous data but listed separately. The defragmenting mimics what the SDK was doing and I noticed it during JPEG handling. Regardless, I protected my changes with a `#[cfg(feature = "otf")]` to help keep up with what was original and what I added.






# Verification Instructions


Use the c2patool changes from https://github.com/Monotype/c2patool/pull/9  to verify font hashing has improved:

- Should be able to sign a font and muck with the font and have it where the validation states it is not valid
    - > NOTE: this was not the case before cause the SDK would glob all data in between CAI records as one, assuming the jumbf blocks were all next to each other

Since I mucked with code outside of the `otf_io` handler, verify you can still sign/validate a JPEG and/or PNG.

> Actively maintained by the @Monotype/driverpdldev team.
